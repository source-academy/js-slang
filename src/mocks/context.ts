import * as es from 'estree'

import createContext from '../createContext'
import Closure from '../interpreter/closure'
import { Context, Environment, Variant } from '../types'

export function mockContext(chapter = 1, variant: Variant = 'default'): Context {
  return createContext(chapter, variant)
}

export function mockRuntimeContext(): Context {
  const context = createContext()
  context.runtime = {
    break: false,
    debuggerOn: true,
    isRunning: true,
    environments: [],
    nodes: [
      {
        type: 'Literal',
        loc: {
          start: { line: 1, column: 0 },
          end: { line: 1, column: 1 }
        },
        value: 0,
        raw: '0',
        range: [0, 1]
      }
    ]
  }
  return context
}

export function mockClosure(): Closure {
  return new Closure(
    {
      type: 'FunctionExpression',
      loc: null,
      id: null,
      params: [],
      body: {
        type: 'BlockStatement',
        body: []
      }
    } as es.FunctionExpression,
    {} as Environment,
    {} as Context
  )
}
