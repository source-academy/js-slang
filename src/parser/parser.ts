/* tslint:disable:max-classes-per-file */
import {
  Options as AcornOptions,
  parse as acornParse,
  parseExpressionAt as acornParseAt,
  Position
} from 'acorn'
import { parse as acornLooseParse } from 'acorn-loose'
import { ancestor, AncestorWalkerFn } from 'acorn-walk/dist/walk'
import * as es from 'estree'
import { Context, ErrorSeverity, ErrorType, Rule, SourceError } from '../types'
import { stripIndent } from '../utils/formatters'
import rules from './rules'
import syntaxBlacklist from './syntaxBlacklist'

export class DisallowedConstructError implements SourceError {
  public type = ErrorType.SYNTAX
  public severity = ErrorSeverity.ERROR
  public nodeType: string

  constructor(public node: es.Node) {
    this.nodeType = this.formatNodeType(this.node.type)
  }

  get location() {
    return this.node.loc!
  }

  public explain() {
    return `${this.nodeType} are not allowed`
  }

  public elaborate() {
    return stripIndent`
      You are trying to use ${this.nodeType}, which is not allowed (yet).
    `
  }

  /**
   * Converts estree node.type into english
   * e.g. ThisExpression -> 'this' expressions
   *      Property -> Properties
   *      EmptyStatement -> Empty Statements
   */
  private formatNodeType(nodeType: string) {
    switch (nodeType) {
      case 'ThisExpression':
        return "'this' expressions"
      case 'Property':
        return 'Properties'
      default: {
        const words = nodeType.split(/(?=[A-Z])/)
        return words.map((word, i) => (i === 0 ? word : word.toLowerCase())).join(' ') + 's'
      }
    }
  }
}

export class FatalSyntaxError implements SourceError {
  public type = ErrorType.SYNTAX
  public severity = ErrorSeverity.ERROR
  public constructor(public location: es.SourceLocation, public message: string) {}

  public explain() {
    return this.message
  }

  public elaborate() {
    return 'There is a syntax error in your program'
  }
}

export class MissingSemicolonError implements SourceError {
  public type = ErrorType.SYNTAX
  public severity = ErrorSeverity.ERROR
  public constructor(public location: es.SourceLocation) {}

  public explain() {
    return 'Missing semicolon at the end of statement'
  }

  public elaborate() {
    return 'Every statement must be terminated by a semicolon.'
  }
}

export class TrailingCommaError implements SourceError {
  public type: ErrorType.SYNTAX
  public severity: ErrorSeverity.WARNING
  public constructor(public location: es.SourceLocation) {}

  public explain() {
    return 'Trailing comma'
  }

  public elaborate() {
    return 'Please remove the trailing comma'
  }
}

export function parseAt(source: string, num: number) {
  let theNode: acorn.Node | undefined
  try {
    theNode = acornParseAt(source, num)
  } catch (error) {
    return undefined
  }
  return theNode
}

export function parse(source: string, context: Context, fallbackToLooseParse: boolean = false) {
  let program: es.Program | undefined
  try {
    program = (acornParse(source, createAcornParserOptions(context)) as unknown) as es.Program
    ancestor(program as es.Node, walkers, undefined, context)
  } catch (error) {
    if (error instanceof SyntaxError) {
      // tslint:disable-next-line:no-any
      const loc = (error as any).loc
      const location = {
        start: { line: loc.line, column: loc.column },
        end: { line: loc.line, column: loc.column + 1 }
      }
      context.errors.push(new FatalSyntaxError(location, error.toString()))
    } else {
      throw error
    }
  }
  const hasErrors = context.errors.find(m => m.severity === ErrorSeverity.ERROR)
  if (program && !hasErrors) {
    return program
  } else if (fallbackToLooseParse) {
    return looseParse(source, context)
  } else {
    return undefined
  }
}

const createAcornParserOptions = (context: Context): AcornOptions => ({
  sourceType: 'module',
  ecmaVersion: 6,
  locations: true,
  // tslint:disable-next-line:no-any
  onInsertedSemicolon(end: any, loc: any) {
    context.errors.push(
      new MissingSemicolonError({
        end: { line: loc.line, column: loc.column + 1 },
        start: loc
      })
    )
  },
  // tslint:disable-next-line:no-any
  onTrailingComma(end: any, loc: Position) {
    context.errors.push(
      new TrailingCommaError({
        end: { line: loc.line, column: loc.column + 1 },
        start: loc
      })
    )
  }
})

// Names-extractor needs comments
export function parseForNames(source: string): [es.Program, acorn.Comment[]] {
  let comments: acorn.Comment[] = []
  const options: AcornOptions = {
    sourceType: 'module',
    ecmaVersion: 6,
    locations: true,
    onComment: comments
  }
  let program: es.Program | undefined
  try {
    program = (acornParse(source, options) as unknown) as es.Program
  } catch {
    comments = []
    program = acornLooseParse(source, options)
  }

  return [program, comments]
}

export function looseParse(source: string, context: Context) {
  const program = (acornLooseParse(
    source,
    createAcornParserOptions(context)
  ) as unknown) as es.Program
  return program
}

function createWalkers(
  allowedSyntaxes: { [nodeName: string]: number },
  parserRules: Rule<es.Node>[]
) {
  const newWalkers = new Map<string, AncestorWalkerFn<Context>>()
  const visitedNodes = new Set<es.Node>()

  // Provide callbacks checking for disallowed syntaxes, such as case, switch...
  const syntaxPairs = Object.entries(allowedSyntaxes)
  syntaxPairs.map(pair => {
    const syntax = pair[0]
    newWalkers.set(syntax, (node: es.Node, context: Context, ancestors: [es.Node]) => {
      if (!visitedNodes.has(node)) {
        visitedNodes.add(node)

        if (context.chapter < allowedSyntaxes[node.type]) {
          context.errors.push(new DisallowedConstructError(node))
        }
      }
    })
  })

  // Provide callbacks checking for rule violations, e.g. no block arrow funcs, non-empty lists...
  parserRules.forEach(rule => {
    const checkers = rule.checkers
    const syntaxCheckerPair = Object.entries(checkers)
    syntaxCheckerPair.forEach(pair => {
      const syntax = pair[0]
      const checker = pair[1]
      const oldCheck = newWalkers.get(syntax)!
      const newCheck = (node: es.Node, context: Context, ancestors: es.Node[]) => {
        if (typeof rule.disableOn !== 'undefined' && context.chapter >= rule.disableOn) {
          return
        }
        const errors = checker(node, ancestors)
        errors.forEach(e => context.errors.push(e))
      }
      newWalkers.set(syntax, (node: es.Node, context: Context<any>, ancestors) => {
        oldCheck(node, context, ancestors)
        newCheck(node, context, ancestors)
      })
    })
  })

  return mapToObj(newWalkers)
}

const mapToObj = (map: Map<string, any>) =>
  Array.from(map).reduce((obj, [k, v]) => Object.assign(obj, { [k]: v }), {})

const walkers: { [name: string]: AncestorWalkerFn<Context> } = createWalkers(syntaxBlacklist, rules)
