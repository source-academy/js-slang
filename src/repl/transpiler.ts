#!/usr/bin/env node
import { createContext, parseError } from '../index'
import { Variant } from '../types'
import { sourceLanguages } from '../constants'
import { parse } from '../parser/parser'
import { transpile } from '../transpiler/transpiler'
import { transpileToGPU } from '../gpu/gpu'
import { transpileToLazy } from '../lazy/lazy'
import { validateAndAnnotate } from '../validator/validator'
import { Program } from 'estree'
import { generate } from 'astring'

function transpileCode(chapter = 1, variant: Variant = 'default', code = '', pretranspile = false) {
  // use defaults for everything
  const context = createContext(chapter, variant, undefined, undefined)
  const program = parse(code, context)
  if (program === undefined) {
    throw Error(parseError(context.errors, true))
  }
  validateAndAnnotate(program as Program, context)
  switch (variant) {
    case 'gpu':
      transpileToGPU(program)
      break
    case 'lazy':
      transpileToLazy(program)
      break
  }
  if (pretranspile) {
    return generate(program)
  } else {
    return transpile(program as Program, context).transpiled
  }
}

/**
 * Returns true iff the given chapter and variant combination is supported.
 */
function validChapterVariant(chapter: any, variant: any) {
  for (const lang of sourceLanguages) {
    if (lang.chapter === chapter && lang.variant === variant) return true
  }

  return false
}

function main() {
  const opt = require('node-getopt')
    .create([
      ['c', 'chapter=CHAPTER', 'set the Source chapter number (i.e., 1-4)', '1'],
      [
        'p',
        'pretranspile',
        "only pretranspile (e.g. GPU -> Source) and don't perform Source -> JS transpilation"
      ],
      [
        'v',
        'variant=VARIANT',
        'set the Source variant (i.e., default, interpreter, substituter, lazy, non-det, concurrent, wasm, gpu)',
        'default'
      ],
      ['h', 'help', 'display this help']
    ])
    .bindHelp()
    .setHelp('Usage: js-slang-transpiler [OPTION]\n\n[[OPTIONS]]')
    .parseSystem()

  const pretranspile = opt.options.pretranspile
  const variant = opt.options.variant
  const chapter = parseInt(opt.options.chapter, 10)
  const valid = validChapterVariant(chapter, variant)
  if (!valid || !(variant === 'default' || variant === 'lazy' || variant === 'gpu')) {
    throw new Error(
      'The chapter and variant combination provided is unsupported. Use the -h option to view valid chapters and variants.'
    )
  }

  const chunks: Buffer[] = []
  process.stdin.on('data', chunk => {
    chunks.push(chunk)
  })
  process.stdin.on('end', () => {
    const code = Buffer.concat(chunks).toString('utf-8')
    const transpiled = transpileCode(chapter, variant, code, pretranspile)
    process.stdout.write(transpiled)
  })
}

main()
