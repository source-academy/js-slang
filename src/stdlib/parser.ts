import { oneLine } from 'common-tags'
import * as es from 'estree'

import { parse as sourceParse } from '../parser'
import { Context, Value } from '../types'
import { vector_to_list } from './list'

class ParseError extends Error {
  constructor(message: string) {
    super(message)
    this.name = 'ParseError'
  }
}

function unreachable() {
  // tslint:disable-next-line:no-console
  console.error(oneLine`
    UNREACHABLE CODE REACHED!
    Please file an issue at
    https://github.com/source-academy/js-slang/issues
    if you see this.
  `)
}

type ASTTransformers = Map<string, (node: es.Node) => Value>

let transformers: ASTTransformers
transformers = new Map([
  [
    'Program',
    (node: es.Node) => {
      node = node as es.Program
      return vector_to_list(node.body.map(transform))
    }
  ],

  [
    'BlockStatement',
    (node: es.Node) => {
      node = node as es.BlockStatement
      return {
        tag: 'block',
        body: vector_to_list(node.body.map(transform))
      }
    }
  ],

  [
    'ExpressionStatement',
    (node: es.Node) => {
      node = node as es.ExpressionStatement
      return transform(node.expression)
    }
  ],

  [
    'IfStatement',
    (node: es.Node) => {
      node = node as es.IfStatement
      return {
        tag: 'conditional_statement',
        predicate: transform(node.test),
        consequent: transform(node.consequent),
        alternative: transform(node.alternate as es.Statement)
      }
    }
  ],

  [
    'FunctionDeclaration',
    (node: es.Node) => {
      node = node as es.FunctionDeclaration
      return {
        tag: 'constant_declaration',
        name: transform(node.id as es.Identifier),
        value: {
          tag: 'function_definition',
          parameters: vector_to_list(node.params.map(transform)),
          body: vector_to_list(node.body.body.map(transform))
        }
      }
    }
  ],

  [
    'VariableDeclaration',
    (node: es.Node) => {
      node = node as es.VariableDeclaration
      if (node.kind === 'let') {
        return {
          tag: 'variable_declaration',
          name: transform(node.declarations[0].id),
          value: transform(node.declarations[0].init as es.Expression)
        }
      } else if (node.kind === 'const') {
        return {
          tag: 'constant_declaration',
          name: transform(node.declarations[0].id),
          value: transform(node.declarations[0].init as es.Expression)
        }
      } else {
        unreachable()
        throw new ParseError('Invalid declaration kind')
      }
    }
  ],

  [
    'ReturnStatement',
    (node: es.Node) => {
      node = node as es.ReturnStatement
      return {
        tag: 'return_statement',
        expression: transform(node.argument as es.Expression)
      }
    }
  ],

  [
    'CallExpression',
    (node: es.Node) => {
      node = node as es.CallExpression
      return {
        tag: 'application',
        operator: transform(node.callee),
        operands: vector_to_list(node.arguments.map(transform))
      }
    }
  ],

  [
    'UnaryExpression',
    (node: es.Node) => {
      node = node as es.UnaryExpression
      const loc = node.loc as es.SourceLocation
      return {
        tag: 'application',
        operator: {
          tag: 'name',
          name: node.operator,
          loc: {
            start: loc.start,
            end: { line: loc.start.line, column: loc.start.column + 1 }
          } as es.SourceLocation
        },
        operands: vector_to_list([transform(node.argument)])
      }
    }
  ],

  [
    'BinaryExpression',
    (node: es.Node) => {
      node = node as es.BinaryExpression
      const loc = node.right.loc as es.SourceLocation
      return {
        tag: 'application',
        operator: {
          tag: 'name',
          name: node.operator,
          loc: {
            start: { line: loc.start.line, column: loc.start.column - 1 },
            end: { line: loc.start.line, column: loc.start.column }
          } as es.SourceLocation
        },
        operands: vector_to_list([transform(node.left), transform(node.right)])
      }
    }
  ],

  [
    'LogicalExpression',
    (node: es.Node) => {
      node = node as es.LogicalExpression
      const loc = node.right.loc as es.SourceLocation
      return {
        tag: 'boolean_operation',
        operator: {
          tag: 'name',
          name: node.operator,
          loc: {
            start: { line: loc.start.line, column: loc.start.column - 1 },
            end: { line: loc.start.line, column: loc.start.column }
          } as es.SourceLocation
        },
        operands: vector_to_list([transform(node.left), transform(node.right)])
      }
    }
  ],

  [
    'ConditionalExpression',
    (node: es.Node) => {
      node = node as es.ConditionalExpression
      return {
        tag: 'conditional_expression',
        predicate: transform(node.test),
        consequent: transform(node.consequent),
        alternative: transform(node.alternate)
      }
    }
  ],

  [
    'ArrowFunctionExpression',
    (node: es.Node) => {
      node = node as es.ArrowFunctionExpression
      return {
        tag: 'function_definition',
        parameters: vector_to_list(node.params.map(transform)),
        body: {
          tag: 'return_statement',
          expression: transform(node.body as es.Expression),
          loc: node.body.loc
        }
      }
    }
  ],

  [
    'Identifier',
    (node: es.Node) => {
      node = node as es.Identifier
      return {
        tag: 'name',
        name: node.name
      }
    }
  ],

  [
    'Literal',
    (node: es.Node) => {
      node = node as es.Literal
      return node.value
    }
  ],

  [
    'ArrayExpression',
    (node: es.Node) => {
      node = node as es.ArrayExpression
      if (node.elements.length === 0) {
        return {
          tag: 'empty_list'
        }
      } else {
        return {
          tag: 'array_expression',
          elements: vector_to_list(node.elements.map(transform))
        }
      }
    }
  ],

  [
    'AssignmentExpression',
    (node: es.Node) => {
      node = node as es.AssignmentExpression
      if (node.left.type === 'Identifier') {
        return {
          tag: 'assignment',
          name: transform(node.left as es.Identifier),
          value: transform(node.right)
        }
      } else if (node.left.type === 'MemberExpression') {
        return {
          tag: 'property_assignment',
          object: transform(node.left as es.Expression),
          value: transform(node.right)
        }
      } else {
        unreachable()
        throw new ParseError('Invalid assignment')
      }
    }
  ],

  [
    'ForStatement',
    (node: es.Node) => {
      node = node as es.ForStatement
      return {
        tag: 'for_loop',
        initialiser: transform(node.init as es.VariableDeclaration | es.Expression),
        predicate: transform(node.test as es.Expression),
        finaliser: transform(node.update as es.Expression),
        statements: transform(node.body)
      }
    }
  ],

  [
    'WhileStatement',
    (node: es.Node) => {
      node = node as es.WhileStatement
      return {
        tag: 'while_loop',
        predicate: transform(node.test),
        statements: transform(node.body)
      }
    }
  ],

  [
    'BreakStatement',
    (node: es.Node) => {
      node = node as es.BreakStatement
      return {
        tag: 'break_statement'
      }
    }
  ],

  [
    'ContinueStatement',
    (node: es.Node) => {
      node = node as es.ContinueStatement
      return {
        tag: 'continue_statement'
      }
    }
  ],

  [
    'ObjectExpression',
    (node: es.Node) => {
      node = node as es.ObjectExpression
      return {
        tag: 'object_expression',
        pairs: vector_to_list(node.properties.map(transform))
      }
    }
  ],

  [
    'MemberExpression',
    (node: es.Node) => {
      node = node as es.MemberExpression
      if (node.computed) {
        return {
          tag: 'property_access',
          object: transform(node.object),
          property: transform(node.property)
        }
      } else {
        const prop = node.property as es.Identifier
        return {
          tag: 'property_access',
          object: transform(node.object),
          property: prop.name
        }
      }
    }
  ],

  [
    'Property',
    (node: es.Node) => {
      node = node as es.Property
      if (node.key.type === 'Literal') {
        return [node.key.value, transform(node.value)]
      } else if (node.key.type === 'Identifier') {
        return [node.key.name, transform(node.value)]
      } else {
        unreachable()
        throw new ParseError('Invalid property key type')
      }
    }
  ]
])

function transform(node: es.Node) {
  if (transformers.has(node.type)) {
    const transformer = transformers.get(node.type) as (n: es.Node) => Value
    const transformed = transformer(node)
    // Attach location information
    if (
      transformed !== null &&
      transformed !== undefined &&
      typeof transformed === 'object' &&
      transformed.tag !== undefined
    ) {
      transformed.loc = node.loc
    }
    return transformed
  } else {
    unreachable()
    throw new ParseError('Cannot transform unknown type: ' + node.type)
  }
}

export function parse(x: string, context: Context): Value {
  let program
  program = sourceParse(x, context)
  if (context.errors.length > 0) {
    throw new ParseError(context.errors[0].explain())
  }

  if (program !== undefined) {
    return transform(program)
  } else {
    unreachable()
    throw new ParseError('Invalid parse')
  }
}
