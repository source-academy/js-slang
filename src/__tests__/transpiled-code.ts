import { stripIndent } from 'common-tags'
import { mockContext } from '../mocks/context'
import { parse } from '../parser'
import { transpile } from '../transpiler'

/*  DO NOT HAVE 'native[<digit>]' AS A SUBSTRING IN CODE STRINGS ANYWHERE IN THIS FILE!
 *  Some code here have a redundant '1;' as the last statement to prevent the
 *  code being tested from being transformed into eval.
 *  Check for variables being stored back by looking at all the tests.
 */
test('builtins do get prepended', () => {
  const code = 'ensure_builtins;'
  const context = mockContext(4)
  const transpiled = transpile(parse(code, context)!, context.contextId).transpiled
  // replace native[<number>] as they may be inconsistent
  const replacedNative = transpiled.replace(/native\[\d+]/g, 'native')
  // replace the line hiding globals as they may differ between environments
  const replacedGlobalsLine = replacedNative.replace(/\n\(\(.*\)/, '\n(( <globals redacted> )')
  expect({ code, transpiled: replacedGlobalsLine }).toMatchSnapshot()
})

test('Ensure no name clashes', () => {
  const code = stripIndent`
    boolOrErr[123] = 1;
    function f(callIfFuncAndRightArgs) {
      let wrap = 2;
      wrap0;wrap1;wrap2;wrap3;wrap4;wrap5;wrap6;wrap7;wrap8;wrap9;
    }
    native;
  `
  const context = mockContext(4)
  const transpiled = transpile(parse(code, context)!, context.contextId).transpiled
  const replacedNative = transpiled.replace(/native0\[\d+]/g, 'native')
  const replacedGlobalsLine = replacedNative.replace(/\n\(\(.*\)/, '\n(( <globals redacted> )')
  expect(replacedGlobalsLine).toMatchSnapshot()
})
