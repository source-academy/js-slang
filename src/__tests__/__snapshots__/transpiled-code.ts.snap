// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Ensure no name clashes 1`] = `
"const native0 = $$NATIVE_STORAGE;
const boolOrErr0 = native.operators.get(\\"itselfIfBooleanElseError\\");
const callIfFuncAndRightArgs0 = native.operators.get(\\"callIfFunctionAndRightArgumentsElseError\\");
const wrap90 = native.properTailCalls.wrap;
(( <globals redacted> ) => {
  return (() => {
    const runtime = native.builtins.get(\\"runtime\\");
    const display = native.builtins.get(\\"display\\");
    const raw_display = native.builtins.get(\\"raw_display\\");
    const stringify = native.builtins.get(\\"stringify\\");
    const error = native.builtins.get(\\"error\\");
    const prompt = native.builtins.get(\\"prompt\\");
    const is_number = native.builtins.get(\\"is_number\\");
    const is_string = native.builtins.get(\\"is_string\\");
    const is_function = native.builtins.get(\\"is_function\\");
    const is_boolean = native.builtins.get(\\"is_boolean\\");
    const is_undefined = native.builtins.get(\\"is_undefined\\");
    const parse_int = native.builtins.get(\\"parse_int\\");
    const undefined = native.builtins.get(\\"undefined\\");
    const NaN = native.builtins.get(\\"NaN\\");
    const Infinity = native.builtins.get(\\"Infinity\\");
    const math_abs = native.builtins.get(\\"math_abs\\");
    const math_acos = native.builtins.get(\\"math_acos\\");
    const math_acosh = native.builtins.get(\\"math_acosh\\");
    const math_asin = native.builtins.get(\\"math_asin\\");
    const math_asinh = native.builtins.get(\\"math_asinh\\");
    const math_atan = native.builtins.get(\\"math_atan\\");
    const math_atanh = native.builtins.get(\\"math_atanh\\");
    const math_atan2 = native.builtins.get(\\"math_atan2\\");
    const math_ceil = native.builtins.get(\\"math_ceil\\");
    const math_cbrt = native.builtins.get(\\"math_cbrt\\");
    const math_expm1 = native.builtins.get(\\"math_expm1\\");
    const math_clz32 = native.builtins.get(\\"math_clz32\\");
    const math_cos = native.builtins.get(\\"math_cos\\");
    const math_cosh = native.builtins.get(\\"math_cosh\\");
    const math_exp = native.builtins.get(\\"math_exp\\");
    const math_floor = native.builtins.get(\\"math_floor\\");
    const math_fround = native.builtins.get(\\"math_fround\\");
    const math_hypot = native.builtins.get(\\"math_hypot\\");
    const math_imul = native.builtins.get(\\"math_imul\\");
    const math_log = native.builtins.get(\\"math_log\\");
    const math_log1p = native.builtins.get(\\"math_log1p\\");
    const math_log2 = native.builtins.get(\\"math_log2\\");
    const math_log10 = native.builtins.get(\\"math_log10\\");
    const math_max = native.builtins.get(\\"math_max\\");
    const math_min = native.builtins.get(\\"math_min\\");
    const math_pow = native.builtins.get(\\"math_pow\\");
    const math_random = native.builtins.get(\\"math_random\\");
    const math_round = native.builtins.get(\\"math_round\\");
    const math_sign = native.builtins.get(\\"math_sign\\");
    const math_sin = native.builtins.get(\\"math_sin\\");
    const math_sinh = native.builtins.get(\\"math_sinh\\");
    const math_sqrt = native.builtins.get(\\"math_sqrt\\");
    const math_tan = native.builtins.get(\\"math_tan\\");
    const math_tanh = native.builtins.get(\\"math_tanh\\");
    const math_trunc = native.builtins.get(\\"math_trunc\\");
    const math_E = native.builtins.get(\\"math_E\\");
    const math_LN10 = native.builtins.get(\\"math_LN10\\");
    const math_LN2 = native.builtins.get(\\"math_LN2\\");
    const math_LOG10E = native.builtins.get(\\"math_LOG10E\\");
    const math_LOG2E = native.builtins.get(\\"math_LOG2E\\");
    const math_PI = native.builtins.get(\\"math_PI\\");
    const math_SQRT1_2 = native.builtins.get(\\"math_SQRT1_2\\");
    const math_SQRT2 = native.builtins.get(\\"math_SQRT2\\");
    const pair = native.builtins.get(\\"pair\\");
    const is_pair = native.builtins.get(\\"is_pair\\");
    const head = native.builtins.get(\\"head\\");
    const tail = native.builtins.get(\\"tail\\");
    const is_null = native.builtins.get(\\"is_null\\");
    const is_list = native.builtins.get(\\"is_list\\");
    const list = native.builtins.get(\\"list\\");
    const length = native.builtins.get(\\"length\\");
    const map = native.builtins.get(\\"map\\");
    const build_list = native.builtins.get(\\"build_list\\");
    const for_each = native.builtins.get(\\"for_each\\");
    const list_to_string = native.builtins.get(\\"list_to_string\\");
    const reverse = native.builtins.get(\\"reverse\\");
    const append = native.builtins.get(\\"append\\");
    const member = native.builtins.get(\\"member\\");
    const remove = native.builtins.get(\\"remove\\");
    const remove_all = native.builtins.get(\\"remove_all\\");
    const filter = native.builtins.get(\\"filter\\");
    const enum_list = native.builtins.get(\\"enum_list\\");
    const list_ref = native.builtins.get(\\"list_ref\\");
    const accumulate = native.builtins.get(\\"accumulate\\");
    const equal = native.builtins.get(\\"equal\\");
    const draw_data = native.builtins.get(\\"draw_data\\");
    const set_head = native.builtins.get(\\"set_head\\");
    const set_tail = native.builtins.get(\\"set_tail\\");
    const array_length = native.builtins.get(\\"array_length\\");
    const is_array = native.builtins.get(\\"is_array\\");
    const parse = native.builtins.get(\\"parse\\");
    const apply_in_underlying_javascript = native.builtins.get(\\"apply_in_underlying_javascript\\");
    boolOrErr[123] = 1;
    const f = wrap90(callIfFuncAndRightArgs => {
      let wrap = 2;
      wrap0;
      wrap1;
      wrap2;
      wrap3;
      wrap4;
      wrap5;
      wrap6;
      wrap7;
      wrap8;
      wrap9;
    }, \\"function f(callIfFuncAndRightArgs) {\\\\n  let wrap = 2;\\\\n  wrap0;\\\\n  wrap1;\\\\n  wrap2;\\\\n  wrap3;\\\\n  wrap4;\\\\n  wrap5;\\\\n  wrap6;\\\\n  wrap7;\\\\n  wrap8;\\\\n  wrap9;\\\\n}\\");
    const lastStatementResult = eval(\\"native;\\");
    native.globals.set(\\"f\\", {
      kind: \\"const\\",
      value: f
    });
    return lastStatementResult;
  })();
})();
"
`;

exports[`builtins do get prepended 1`] = `
Object {
  "code": "ensure_builtins;",
  "transpiled": "const native = $$NATIVE_STORAGE;
const boolOrErr = native.operators.get(\\"itselfIfBooleanElseError\\");
const callIfFuncAndRightArgs = native.operators.get(\\"callIfFunctionAndRightArgumentsElseError\\");
const wrap = native.properTailCalls.wrap;
(( <globals redacted> ) => {
  return (() => {
    const runtime = native.builtins.get(\\"runtime\\");
    const display = native.builtins.get(\\"display\\");
    const raw_display = native.builtins.get(\\"raw_display\\");
    const stringify = native.builtins.get(\\"stringify\\");
    const error = native.builtins.get(\\"error\\");
    const prompt = native.builtins.get(\\"prompt\\");
    const is_number = native.builtins.get(\\"is_number\\");
    const is_string = native.builtins.get(\\"is_string\\");
    const is_function = native.builtins.get(\\"is_function\\");
    const is_boolean = native.builtins.get(\\"is_boolean\\");
    const is_undefined = native.builtins.get(\\"is_undefined\\");
    const parse_int = native.builtins.get(\\"parse_int\\");
    const undefined = native.builtins.get(\\"undefined\\");
    const NaN = native.builtins.get(\\"NaN\\");
    const Infinity = native.builtins.get(\\"Infinity\\");
    const math_abs = native.builtins.get(\\"math_abs\\");
    const math_acos = native.builtins.get(\\"math_acos\\");
    const math_acosh = native.builtins.get(\\"math_acosh\\");
    const math_asin = native.builtins.get(\\"math_asin\\");
    const math_asinh = native.builtins.get(\\"math_asinh\\");
    const math_atan = native.builtins.get(\\"math_atan\\");
    const math_atanh = native.builtins.get(\\"math_atanh\\");
    const math_atan2 = native.builtins.get(\\"math_atan2\\");
    const math_ceil = native.builtins.get(\\"math_ceil\\");
    const math_cbrt = native.builtins.get(\\"math_cbrt\\");
    const math_expm1 = native.builtins.get(\\"math_expm1\\");
    const math_clz32 = native.builtins.get(\\"math_clz32\\");
    const math_cos = native.builtins.get(\\"math_cos\\");
    const math_cosh = native.builtins.get(\\"math_cosh\\");
    const math_exp = native.builtins.get(\\"math_exp\\");
    const math_floor = native.builtins.get(\\"math_floor\\");
    const math_fround = native.builtins.get(\\"math_fround\\");
    const math_hypot = native.builtins.get(\\"math_hypot\\");
    const math_imul = native.builtins.get(\\"math_imul\\");
    const math_log = native.builtins.get(\\"math_log\\");
    const math_log1p = native.builtins.get(\\"math_log1p\\");
    const math_log2 = native.builtins.get(\\"math_log2\\");
    const math_log10 = native.builtins.get(\\"math_log10\\");
    const math_max = native.builtins.get(\\"math_max\\");
    const math_min = native.builtins.get(\\"math_min\\");
    const math_pow = native.builtins.get(\\"math_pow\\");
    const math_random = native.builtins.get(\\"math_random\\");
    const math_round = native.builtins.get(\\"math_round\\");
    const math_sign = native.builtins.get(\\"math_sign\\");
    const math_sin = native.builtins.get(\\"math_sin\\");
    const math_sinh = native.builtins.get(\\"math_sinh\\");
    const math_sqrt = native.builtins.get(\\"math_sqrt\\");
    const math_tan = native.builtins.get(\\"math_tan\\");
    const math_tanh = native.builtins.get(\\"math_tanh\\");
    const math_trunc = native.builtins.get(\\"math_trunc\\");
    const math_E = native.builtins.get(\\"math_E\\");
    const math_LN10 = native.builtins.get(\\"math_LN10\\");
    const math_LN2 = native.builtins.get(\\"math_LN2\\");
    const math_LOG10E = native.builtins.get(\\"math_LOG10E\\");
    const math_LOG2E = native.builtins.get(\\"math_LOG2E\\");
    const math_PI = native.builtins.get(\\"math_PI\\");
    const math_SQRT1_2 = native.builtins.get(\\"math_SQRT1_2\\");
    const math_SQRT2 = native.builtins.get(\\"math_SQRT2\\");
    const pair = native.builtins.get(\\"pair\\");
    const is_pair = native.builtins.get(\\"is_pair\\");
    const head = native.builtins.get(\\"head\\");
    const tail = native.builtins.get(\\"tail\\");
    const is_null = native.builtins.get(\\"is_null\\");
    const is_list = native.builtins.get(\\"is_list\\");
    const list = native.builtins.get(\\"list\\");
    const length = native.builtins.get(\\"length\\");
    const map = native.builtins.get(\\"map\\");
    const build_list = native.builtins.get(\\"build_list\\");
    const for_each = native.builtins.get(\\"for_each\\");
    const list_to_string = native.builtins.get(\\"list_to_string\\");
    const reverse = native.builtins.get(\\"reverse\\");
    const append = native.builtins.get(\\"append\\");
    const member = native.builtins.get(\\"member\\");
    const remove = native.builtins.get(\\"remove\\");
    const remove_all = native.builtins.get(\\"remove_all\\");
    const filter = native.builtins.get(\\"filter\\");
    const enum_list = native.builtins.get(\\"enum_list\\");
    const list_ref = native.builtins.get(\\"list_ref\\");
    const accumulate = native.builtins.get(\\"accumulate\\");
    const equal = native.builtins.get(\\"equal\\");
    const draw_data = native.builtins.get(\\"draw_data\\");
    const set_head = native.builtins.get(\\"set_head\\");
    const set_tail = native.builtins.get(\\"set_tail\\");
    const array_length = native.builtins.get(\\"array_length\\");
    const is_array = native.builtins.get(\\"is_array\\");
    const parse = native.builtins.get(\\"parse\\");
    const apply_in_underlying_javascript = native.builtins.get(\\"apply_in_underlying_javascript\\");
    const lastStatementResult = eval(\\"ensure_builtins;\\");
    return lastStatementResult;
  })();
})();
",
}
`;
